### ISSUE TITLE
multiple lifted nodes from identical code fails

### ISSUE THREAD

**Issue by paciorek** [nimble-dev/nimble#337]
I haven't investigated at all as I'm bogged down in some PalEON coding that revealed this:

```
code = nimbleCode({
for(j in 1:3) {  
      omega[j] ~ dexp(lambda[j]*lambda[j]/2)
      lambda[j] ~ dgamma(1,1)
  }
  for(j in (order+1):J) {  
      omega[j] ~ dexp(lambda[j]*lambda[j]/2)
      lambda[j] ~ dgamma(1, rate = sigma) 
  }
})

model = nimbleModel(code, constants = list(J=100, order =3))
model$getDependencies('lambda[4]')
[1] "lambda[4]"
```
Note that the dependencies of lambda[4] are incorrect. 

Things are fine if I use a distinct index:
```
code = nimbleCode({
for(j in 1:3) {  
      omega[j] ~ dexp(lambda[j]*lambda[j]/2)
      lambda[j] ~ dgamma(1,1)
  }
  for(k in (order+1):J) {  
      omega[k] ~ dexp(lambda[k]*lambda[k]/2)
      lambda[k] ~ dgamma(1, rate = sigma) 
  }
})

model = nimbleModel(code, constants = list(J=100, order =3))
model$getDependencies('lambda[4]')
[1] "lambda[4]"                                         
[2] "lifted_lambda_oBk_cB_times_lambda_oBk_cB_over_2[4]"
[3] "omega[4]"                                          
```

**Comment by perrydv** [comment]
This occurs because the two `omega[j]` declarations create a lifted declaration with *identical* code.  But lifted declarations are meant to be unique by name (a mangling of the expression), and a declaration can have only one `context` (for-loop info).  This works well if say `rate = sigma` appears in multiple different for-loops, because only a *single* `d1_over_sigma <- 1 / sigma` declaration is lifted.  But in the example here, the second declaration that would lift `lambda[j] * lambda[j] / 2` does not do so because it appears identical to the first one, so it is not lifted again. The first one appears in the `j in 1:3` context, so the next context is effectively ignored for the lifted declaration 

The simplest rough solution would be to embed the existing `contextID` (unique for each for-loop situation) into the lifted node name.  The would solve the current problem by ensuring that lifted nodes are never shared across contexts.  But it would result in duplicate lifts, something like `d1_over_sigma_context1 <- 1/ sigma` and `d1_over_sigma_context2 <- 1/sigma`, which might not be a big deal except for computationally heavy lifted nodes such as with `chol()`.  

A more in-depth solution would be to add a new stage in processing that carefully inspects the sets of lifted nodes and their contexts and rarefies or merges them as needed.  Non-trivial cases could involve partially indexed (i.e. within a for-loop with i and j, but with lifted node only using i *or* j) cases with overlapping index ranges in different contexts.  We could either create new `context` entries just for lifted nodes when necessary, or when unrolling indices in `modelDef::genNodeInfo3` we could combine from identical declarations in multiple contexts, perhaps explicitly marking lifted declarations (which at the moment are handled as regular declarations at every step after being lifted).

**Comment by perrydv** [comment]
@paciorek, here is my suggested solution: When lifting an expression to its own BUGS statement, we can inspect whether it contains any indices of any for loops it's inside of.  If it does not use any indices, that should guarantee it can use the same lifted node as created by an identical expression elsewhere (if one exists), so we can use it.  If it does use any for-loop indices, we'll simply append a unique integer (or other label) to the end of the lifted node name.  This should ensure that lifted nodes from identical, index-involving expressions are treated as coming from separate declarations.  This seems necessary because the indexing contexts they are in could be overlapping, non-contiguous, or otherwise arbitrary.  It will result in some wasted memory allocation (in the filed example, the lifted node from the second omega[j] declaration will have unused elements from 1 to `order`) but that seems a minor cost to gain a quick fix to the bug.

**Comment by paciorek** [comment]
ok, sounds good to me. I did not parse all of the code in the diff of the commit...

**Comment by perrydv** [comment]
Fixed by PR #350
